{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploratory Analysis of scRNA Datasets & GPU Scaling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.\n",
    "\n",
    "NVIDIA CORPORATION and its licensors retain all intellectual property\n",
    "and proprietary rights in and to this software, related documentation\n",
    "and any modifications thereto.  Any use, reproduction, disclosure or\n",
    "distribution of this software and related documentation without an express\n",
    "license agreement from NVIDIA CORPORATION is strictly prohibited."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook demonstrates an exploratory single-cell RNA analysis workflow that begins with preprocessing a count matrix of size `(n_gene, n_cell)` and results in a visualization of the clustered cells for further analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import requirements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "\n",
    "import sys\n",
    "import time\n",
    "import os\n",
    "\n",
    "import cudf\n",
    "\n",
    "\n",
    "import cupy as cp\n",
    "\n",
    "from cuml.decomposition import PCA\n",
    "from cuml.manifold import TSNE\n",
    "from cuml.cluster import KMeans\n",
    "\n",
    "\n",
    "import tables\n",
    "\n",
    "import rapids_scanpy_funcs\n",
    "\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings('ignore', 'Expected column ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/raid/cnolet/miniconda3/envs/genomics2/lib/python3.7/site-packages/distributed/node.py:244: UserWarning: Port 8787 is already in use.\n",
      "Perhaps you already have a cluster running?\n",
      "Hosting the HTTP server on port 41339 instead\n",
      "  http_address[\"port\"], self.http_server.port\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table style=\"border: 2px solid white;\">\n",
       "<tr>\n",
       "<td style=\"vertical-align: top; border: 0px solid white\">\n",
       "<h3 style=\"text-align: left;\">Client</h3>\n",
       "<ul style=\"text-align: left; list-style: none; margin: 0; padding: 0;\">\n",
       "  <li><b>Scheduler: </b>tcp://127.0.0.1:43757</li>\n",
       "  <li><b>Dashboard: </b><a href='http://127.0.0.1:41339/status' target='_blank'>http://127.0.0.1:41339/status</a></li>\n",
       "</ul>\n",
       "</td>\n",
       "<td style=\"vertical-align: top; border: 0px solid white\">\n",
       "<h3 style=\"text-align: left;\">Cluster</h3>\n",
       "<ul style=\"text-align: left; list-style:none; margin: 0; padding: 0;\">\n",
       "  <li><b>Workers: </b>8</li>\n",
       "  <li><b>Cores: </b>8</li>\n",
       "  <li><b>Memory: </b>1.08 TB</li>\n",
       "</ul>\n",
       "</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "<Client: 'tcp://127.0.0.1:43757' processes=8 threads=8, memory=1.08 TB>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dask_cuda import LocalCUDACluster\n",
    "from dask.distributed import Client\n",
    "\n",
    "cluster = LocalCUDACluster()\n",
    "client = Client(cluster)\n",
    "client"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# marker genes\n",
    "RIBO_GENE_PREFIX = \"RPS\" # Prefix for ribosomal genes to regress out\n",
    "markers = [\"ACE2\", \"TMPRSS2\", \"EPCAM\"] # Marker genes for visualization\n",
    "\n",
    "# filtering cells\n",
    "min_genes_per_cell = 200 # Filter out cells with fewer genes than this expressed \n",
    "max_genes_per_cell = 6000 # Filter out cells with more genes than this expressed \n",
    "\n",
    "# filtering genes\n",
    "n_top_genes = 5000 # Number of highly variable genes to retain\n",
    "\n",
    "# PCA\n",
    "n_components = 50 # Number of principal components to compute\n",
    "\n",
    "# t-SNE\n",
    "tsne_n_pcs = 20 # Number of principal components to use for t-SNE\n",
    "\n",
    "# k-means\n",
    "k = 20 # Number of clusters for k-means\n",
    "\n",
    "# KNN\n",
    "n_neighbors = 15 # Number of nearest neighbors for KNN graph\n",
    "knn_n_pcs = 50 # Number of principal components to use for finding nearest neighbors\n",
    "\n",
    "# UMAP\n",
    "umap_min_dist = 0.3 \n",
    "umap_spread = 1.0\n",
    "\n",
    "# Louvain\n",
    "louvain_resolution = 0.4\n",
    "\n",
    "# Gene ranking\n",
    "ranking_n_top_genes = 50 # Number of differential genes to compute for each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "start = time.time()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and Prepare Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We load the sparse count matrix from an `h5ad` file using Scanpy. The sparse count matrix will then be placed on the GPU. \n",
    " \n",
    "To convert your CSV file into an `h5ad` file, see `csv_to_h5ad.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "one_mil_file = \"../1M_neurons_filtered_gene_bc_matrices_h5.h5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "\n",
    "f = h5py.File(one_mil_file, mode=\"r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "dset = f[\"mm10\"]\n",
    "columnptr = dset[\"indptr\"][:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NNZ: 2624828308\n"
     ]
    }
   ],
   "source": [
    "nnz = columnptr[-1]\n",
    "print(\"NNZ: %s\" % nnz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: [  27998 1306127]\n"
     ]
    }
   ],
   "source": [
    "print(\"Shape: %s\" % (dset[\"shape\"][:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Density: 0.0717775259310603\n"
     ]
    }
   ],
   "source": [
    "print(\"Density: %s\" % (dset[\"data\"].shape[0] / np.product(dset[\"shape\"][:])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fed0c3d4750>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYcAAAD4CAYAAAAHHSreAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nO3deXRcV53g8e+vSqqSVLJ2ydq8R3FwjLNgYicsAUJIHGicnm7oJNMTQ2fG+JD0xvR0m+7pPgOHM51mumkIhKTD1k7AhDTLxBB3J8EQJkBsLCeOHW+RLNuSrH3fVSrVb/6oJ1PWWpYlv1p+n3N03nbvq/ueLf3qvrs8UVWMMcaYaB63C2CMMSb+WHAwxhgzhQUHY4wxU1hwMMYYM4UFB2OMMVOkuV2AhVBUVKQrV650uxjGGJNQDh061KGqxdMdS4rgsHLlSqqrq90uhjHGJBQROTfTMXusZIwxZgoLDsYYY6aw4GCMMWYKCw7GGGOmsOBgjDFmCgsOxhhjprDgYIwxZgoLDmaKcFixqdyNSW1JMQjOXL6hYIiPf+sgp9sH6Roc5R1XFfHUA5vcLpYxxiVWczAA7DvRxoEzXdy8ppA7ri3l5ZoODp7tcrtYxhiXWM0hxe0+UA/At/efY0lGGptWFRAaVw6c6eLxl07z9o8VuFxCY4wbrOZgGB0b583WftZX5OIRwZfmYdvNK9l3so03W/vdLp4xxgUWHAwnWvoIhZUNFbkX9t1/8woy0738yy/qXCyZMcYtFhwMRxt7yc1MZ1lB1oV9+QEff/D2ZTx7+DxtfSMuls4Y4wYLDiluODjOm20DrC/PwSNy0bHff1slobDy69OdLpXOGOMWCw4p7kRLH+NhZUNl3pRjbynLIdufRvU567VkTKqx3kop7nTbAAF/GpX5mRftn+jFVJabwU+Pt7GuLLJ936blV7yMxpgrz2oOKe5c1xArCrKQSY+UJiwvzKK1b4Th4PgVLpkxxk0WHFJYW98IXYNBVhZmzZhmZWEABeq7hq5cwYwxrrPgkMKqz3UDsKIwMGOaZflZeATOdQ1eqWIZY+KABYcUVn22mzSPUJaXMWMaX5qHstxMznVazcGYVGLBIYUdOtdFZX4WaZ7Z/xusLMyioWuIUDh8hUpmjHFbTMFBRO4UkVMiUisiO6c5LiLyiHP8iIjcOFdeESkQkRdFpMZZ5kcd2yAir4jIMRE5KiIzf7U18zIUDPFGUx8rZmlvmLC8MEAorDT32GA4Y1LFnMFBRLzAo8AWYB1wr4ism5RsC1Dl/GwHHosh705gn6pWAfucbUQkDfg2sENVrwXeA4zN/xLNdA439DAe1piCw0Sas53W7mBMqoil5nATUKuqdaoaBJ4Gtk5KsxV4UiP2A3kiUjZH3q3ALmd9F3C3s/4B4Iiqvg6gqp2qav0oF9ihs5HG6OUFcweHnIx08jLTOd8zvNjFMsbEiViCQwXQELXd6OyLJc1seZeqajOAsyxx9l8NqIg8LyKvishfTlcoEdkuItUiUt3e3h7DZZho1ee6uXppNlm+2MZBluT4ae8fXeRSGWPiRSzBYbrRUZPfITlTmljyTpYGvBP4z87yd0XktiknUX1CVTeq6sbi4uI5TmmiqSqHG3q4cXn+3IkdJUsyaO8fZTxsrw81JhXEEhwagWVR25VAU4xpZsvb6jx6wlm2RZ3rF6raoapDwF7gRsyCaekboXd4jHXlOTHnKVniJxRWGrutS6sxqSCW4HAQqBKRVSLiA+4B9kxKswe43+m1tBnodR4VzZZ3D7DNWd8GPOusPw9sEJEsp3H6VuD4PK/PTONkc+QFPmuXLok5T8kSPwA1rQOLUiZjTHyZMzioagh4iMgf7RPAM6p6TER2iMgOJ9leoA6oBb4GfHK2vE6eh4HbRaQGuN3ZRlW7gS8QCSyHgVdV9bkFuFbjONkSCQ7XlMZecyheEulNXNNmwcGYVBBTa6Sq7iUSAKL3PR61rsCDseZ19ncCU9oSnGPfJtKd1SyCky19lOVmkJuVHnOeTJ+XnIw0atrstaHGpAIbIZ2CTrX0c01p7I+UJpQsyeC01RyMSQkWHFJMMBSmtm2AtZfwSGlC8RI/NW0DRCqKxphkZsEhxdR1DBAKK28pm0fNIcfPUHCcpl6bRsOYZGfBIcVc6Kk0z8dKADWt1u5gTLKz4JBiTrb0k+4VVhdlX3Leie6stdbuYEzSs+CQYk629LGmOBtf2qX/0wf8aRQEfBYcjEkBsU2sYxLe7gP1ALxW38OqosCF7Ut1VUm2jXUwJgVYzSGFDAfH6R0eY2nO/F+PUVWSTU1rv/VYMibJWXBIIS19kV5GpZcRHFYXZ9M3EqJ7yF6xYUwys+CQQpqc9zHM9s7ouawosBf/GJMKLDikkPM9wyzJSCMnI/ZpMyZbWRQJDucsOBiT1Cw4pJDzPcNU5GVe1jkq87MQgXOdNnW3McnMgkOKGA2N09E/SvllBoeMdC/luZkWHIxJchYcUkRL7wgKl11zAFhRmGVtDsYkOQsOKeK80xh9uTUHiAQHqzkYk9wsOKSI893DZPvTyMm4/HGPKwoDdA0G6Rux7qzGJCsLDiliojFaRC77XCsLIz2W6q32YEzSsuCQAoaCIdoXoDF6worCAGBjHYxJZja3Ugo40dy3YI3Ruw/UMxoaB2DP4Sb6hkMA3Ldp+WWf2xgTP6zmkALeON8HQEX+wtQc/GlelvjT6BwMLsj5jDHxx4JDCjh6vpfAAjVGTyjI9tE5YMHBmGRlwSEFvN7QQ+UCNUZPKAz46RocXbDzGWPiiwWHJNc7PEZN2wDLnAnzFkphto++kRDBUHhBz2uMiQ8xBQcRuVNETolIrYjsnOa4iMgjzvEjInLjXHlFpEBEXhSRGmeZ7+xfKSLDInLY+Xl8IS40VR1p7AFg+QIHh4KAD4CuIXu0ZEwymjM4iIgXeBTYAqwD7hWRdZOSbQGqnJ/twGMx5N0J7FPVKmCfsz3htKpe7/zsmO/Fmcib30SgcoEaoycUOsGhc8AeLRmTjGKpOdwE1KpqnaoGgaeBrZPSbAWe1Ij9QJ6IlM2Rdyuwy1nfBdx9mddipnG4oYerirPJSPcu6HmLsv0A1ihtTJKKJThUAA1R243OvljSzJZ3qao2AzjLkqh0q0TkNRH5hYi8a7pCich2EakWker29vYYLiP1qCqv1Xdzw/K8BT93RrqXgD+NDqs5GJOUYgkO03VxmfwC4ZnSxJJ3smZguareAHwK2C0iOVNOovqEqm5U1Y3FxcVznDI1nescontojBuW5y/K+YuzfRYcjElSsQSHRmBZ1HYl0BRjmtnytjqPnnCWbQCqOqqqnc76IeA0cHUsF2Mu9lpDN8Ci1Bwg8mipwx4rGZOUYgkOB4EqEVklIj7gHmDPpDR7gPudXkubgV7nUdFsefcA25z1bcCzACJS7DRkIyKriTRy1837ClPYa/U9BHxeqkqWLMr5i7L9DIyGGBkbX5TzG2PcM+eQWVUNichDwPOAF/imqh4TkR3O8ceBvcBdQC0wBHx8trzOqR8GnhGRB4B64CPO/ncDnxWREDAO7FDVrgW52hRzuKGHDZV5eD0LN/gtWlF2pMeSPVoyJvnENJ+Cqu4lEgCi9z0eta7Ag7HmdfZ3ArdNs/8HwA9iKZeZ2XBwnONNfWx/9+pF+4yJHksWHIxJPjZCOkntr+skFFY2rS5ctM8oCPgQoL3f2h2MSTY2ZXeS2X2gHoA9r58n3Suc7RjkfPfwonxWmtdDfsB6LBmTjKzmkIRUlVMt/VxVnE26d3H/iYuyfTZK2pgkZMEhCbX1j9I9NMba0inDQxbcRHfWSLOTMSZZWHBIQqda+gG4emn2on9WUbaf4HiY1j6rPRiTTCw4JKFTrf2U5mSQl+Vb9M+a6LFU1zGw6J9ljLlyLDgkmeHgOOc6B1lbujgD3yabGOtwpmPwinyeMebKsOCQZGra+gkrXHOFgkNOZjrpXqGu3YKDMcnEgkOSebW+myX+NCrzF/blPjPxiFCU7beagzFJxoJDEqltG+DN1gE2rS5YtCkzplNowcGYpGPBIYns+vVZvB7hplWLNyp6OkXZPuq7hux90sYkEQsOSaJ3eIwfvNrIdZV5ZPuv7MD34mw/42GloXvoin6uMWbxWHBIEs8cbGAoOM4ta65srQF+2531jDVKG5M0LDgkgbHxMP/667PctKqA8rzMK/75NtbBmORjwSEJ/Oi185zvGWbHrYs3PfdsMn1eCgM+a5Q2JolYcEhwofEwX/15Lesrcnjv2hLXyrGqKGBjHYxJIhYcEtxPjjRztnOIh95bhciV67462aqiAHVWczAmadj7HBLU7gP1hFX50r4alub46RgYvfAuBzesLs7m3w410j8yxpKMdNfKYYxZGFZzSGDHm/po7x/lvWtL8LhYa4BIzQHgbId1ZzUmGVhwSFCqyv+raacg4GN9Ra7bxWF1cSQ4WI8lY5KDBYcEdaZzkMbuYd5VVeR6rQFgRWEWIlijtDFJwoJDgnr5zQ4CPi83Ls93uygA+NO8VOZnWndWY5KEBYcEdLKlj1Ot/dy8pmjR3xF9KVYVZdtjJWOSREx/WUTkThE5JSK1IrJzmuMiIo84x4+IyI1z5RWRAhF5UURqnGX+pHMuF5EBEfmLy7nAZPT1l8+Q7hU2ry5wuygXWV0U4Ez7oL1P2pgkMGdwEBEv8CiwBVgH3Csi6yYl2wJUOT/bgcdiyLsT2KeqVcA+ZzvaPwP/Po9rSmp9I2P85EgT1y/LJ8sXXz2RVxcHGAyO095v75M2JtHFUnO4CahV1TpVDQJPA1snpdkKPKkR+4E8ESmbI+9WYJezvgu4e+JkInI3UAccm+d1Ja09h5sYGQvz9pXx0dYQbaI7qw2GMybxxRIcKoCGqO1GZ18saWbLu1RVmwGcZQmAiASAvwI+M1uhRGS7iFSLSHV7e3sMl5EcvnewgbeU5VDhwgR7c7kQHKzHkjEJL5bgMF0/yckPlWdKE0veyT4D/LOqztqyqapPqOpGVd1YXFw8xymTw7GmXo6e7+UPNla6OlXGTMpzM/GneThjjdLGJLxYHlo3AsuitiuBphjT+GbJ2yoiZara7DyCanP2bwJ+X0Q+D+QBYREZUdWvxHJByeyZgw340jzcfUMFe4+2uF2ci0xM3ZGf5ePlmo4L2/dtWu5msYwx8xRLzeEgUCUiq0TEB9wD7JmUZg9wv9NraTPQ6zwqmi3vHmCbs74NeBZAVd+lqitVdSXwReB/W2CA0dA4P3rtPFvWl5KX5XO7ODMqzPbRMWAN0sYkujlrDqoaEpGHgOcBL/BNVT0mIjuc448De4G7gFpgCPj4bHmdUz8MPCMiDwD1wEcW9MqSzKFz3fSNhPjQhnK3izKr4mw/J5r7GA8rXk/8PfoyxsQmpr6QqrqXSACI3vd41LoCD8aa19nfCdw2x+f+r1jKl8wmHs+8cLwFj0BD15Crs6/OpSjbT1iheyh44Q1xxpjEEz/Da82sTrcNUJmfRUa61+2izKooO/LIq8PGOhiT0Cw4JICRsXEau4dZU5ztdlHmVLQkUluwdgdjEpsFhwRQ1z6IAleVxH9wyPKlkeXz0jEQdLsoxpjLYMEhAdS2D5DuFZYVxN/At+kUZftpt5qDMQnNgkMCON0+wKqiAGmexPjnKsr202nBwZiElhh/bVJY7/AY7f2jCdHeMKEo20ffSIjR0LjbRTHGzJMFhzh3uj0yFUViBYeJRmlrdzAmUVlwiHON3cP40jyU5ma4XZSYWY8lYxKfBYc419I7QmlORly8JzpWhQEfgo11MCaRWXCIY6pKS99wQtUaANK9HvIDPlr7RtwuijFmniw4xLGm3hFGxsKUJVhwACjPy6Sp14KDMYnKgkMcO9ncB0BpTuIFh4rcDLoGg/QOjbldFGPMPFhwiGMnEjg4lDlvqjvW3OtySYwx82HBIY6daOmnIODDH+eT7U2nfCI4nO9zuSTGmPmw4BDHTjb3JWStASDbn0ZuZjrHmqzmYEwisuAQp0bGxjnTMZhwPZWiledm8EaT1RyMSUQWHOLUm639hDUx2xsmlOdlcrp9gKFgyO2iGGMukQWHOHWyuR8gIbuxTijPy0T1tw3rxpjEYcEhTh1v7iPL5yU/4HO7KPM20Sj9hjVKG5NwLDjEqZMtfawtXZJQ02ZMlpORRmHAxxvnrVHamERjwSFOnW4fpCoB3vw2GxHh2opca5Q2JgFZcIhDo6Fx2vtHqcjLcrsol219eQ41rf2MjNm7HYxJJDEFBxG5U0ROiUitiOyc5riIyCPO8SMicuNceUWkQEReFJEaZ5nv7L9JRA47P6+LyO8uxIUmktbeyGymZXmJ2xg9YUNlHqGwctwapY1JKHMGBxHxAo8CW4B1wL0ism5Ssi1AlfOzHXgshrw7gX2qWgXsc7YB3gA2qur1wJ3Av4hI2ryvMAE19Q4DUJ6bGO+Mns11y3IBONLQ43JJjDGXIpaaw01ArarWqWoQeBrYOinNVuBJjdgP5IlI2Rx5twK7nPVdwN0AqjqkqhMd4zMAnee1JaymHic4JEHNoTQng+Ilfo40WqO0MYkkluBQATREbTc6+2JJM1veparaDOAsSyYSicgmETkGHAV2RAWLlNDsTHVdlgQ1BxHhuspcXm+0moMxiSSW4DBdX8rJ3+ZnShNL3qkJVA+o6rXA24FPi8iUr9Aisl1EqkWkur29fa5TJpTzPcMUBHxk+hJvwr3pbKjMo65jkP4Rm77bmEQRS3BoBJZFbVcCTTGmmS1vq/PoCWfZNvmDVfUEMAisn+bYE6q6UVU3FhcXx3AZiaO5ZzihR0ZPtqEyF1U4auMdjEkYsQSHg0CViKwSER9wD7BnUpo9wP1Or6XNQK/zqGi2vHuAbc76NuBZACdtmrO+AlgLnJ3vBSaipp6RC6OLE93uA/XUtA4A8OSvz7H7QD27D9S7XCpjzFzm7AWkqiEReQh4HvAC31TVYyKywzn+OLAXuAuoBYaAj8+W1zn1w8AzIvIAUA98xNn/TmCniIwBYeCTqtqxIFebIJp6h9m8usDtYiyYgD+NgoCPxu4ht4tijIlRTF1EVXUvkQAQve/xqHUFHow1r7O/E7htmv1PAU/FUq5k1D8yRv9IKGlqDhMq8jJp6LLgYEyisBHSceZCT6UkCw6V+Zn0DI8xMJpSHc+MSVgWHOLMeWeMQ0USjHGIVpkfmQrEHi0ZkxhSauRxPJtopP3NmS4ADtR1caplwM0iLaiKvEy8IpzrHOKa0hy3i2OMmYPVHOJM73AQAZZkpLtdlAXlS/NQkZ9JXXvyBDxjkpkFhzjTMzRGTmY6Xk/ivsdhJquLApzvGWY0ZDO0GhPvLDjEmd7hMXIzk6vWMGFVUYCwQn2ntTsYE+8sOMSZnuEx8rKSMzgsL8zCI1DXMeh2UYwxc7DgEEfCqvQlcc3Bn+alIi+TMxYcjIl7FhziyOBoiFBYyUvS4ACwujibxu4hhoI23sGYeGbBIY70DkdmLc3N9LlcksUz0e5w6Fy320UxxszCgkMc6R+JfJvOyUze4ScrCiLtDgfqutwuijFmFhYc4sjE1BLZ/uQNDv70SLvDL2tTai5FYxKOBYc4MugEh0ASBweAtaU5HG7oobVvxO2iGGNmYMEhjgyMhvCneUj3Jvc/y7XlkekzXjje6nJJjDEzSe6/QglmYDSU1I+UJpQs8bOqKMALx1rcLooxZgYWHOLIwGgo6R8pAYgId1xbyiunO+kdsvdKGxOPLDjEkcEUqTkA3HHtUkJh5Wen7NGSMfHIgkMcGRgdT5ngcF1lHktz/Dz/hgUHY+KRBYc4EVZlKEUeKwF4PMIH1pXy0pttDAdtllZj4o0FhzgxFBxHgWy/1+2iXDFb1pcyMhbmZyfb3C6KMWaS1PiamgAuDIBLspf8zGT3gXrCqizJSOPRn9fSOzzGfZuWu10sY4zDag5xYmBkYgBc6tQcPCK8tSKXN1v7GRmzR0vGxBMLDnFiYnR0ti+1KnPXVeYRCivHmvrcLooxJkpMwUFE7hSRUyJSKyI7pzkuIvKIc/yIiNw4V14RKRCRF0WkxlnmO/tvF5FDInLUWb5vIS403qXCvErTqczPpCDg40hjj9tFMcZEmTM4iIgXeBTYAqwD7hWRdZOSbQGqnJ/twGMx5N0J7FPVKmCfsw3QAfyOqr4V2AY8Ne+rSyCDoyE8Ahm+1HmsBJEBcRsqcjndPkDHwKjbxTHGOGKpOdwE1KpqnaoGgaeBrZPSbAWe1Ij9QJ6IlM2Rdyuwy1nfBdwNoKqvqWqTs/8YkCEi/nleX8KYGB3tEXG7KFfchmV5hBX2Hm12uyjGGEcswaECaIjabnT2xZJmtrxLVbUZwFmWTPPZvwe8pqpTvlKKyHYRqRaR6vb29hguI76lyrxK0ynNyWBpjp89h5vmTmyMuSJiCQ7TfZXVGNPEknf6DxW5FvgH4BPTHVfVJ1R1o6puLC4ujuWUcW0whQbATee6yjyqz3VzvmfY7aIYY4gtODQCy6K2K4HJX/FmSjNb3lbn0RPO8sJIKBGpBH4E3K+qp2MoY8JL5ZoDwIbKPAB+/LrVHoyJB7EEh4NAlYisEhEfcA+wZ1KaPcD9Tq+lzUCv86hotrx7iDQ44yyfBRCRPOA54NOq+qvLuLaEkurBoSDg4/plefZoyZg4MWdwUNUQ8BDwPHACeEZVj4nIDhHZ4STbC9QBtcDXgE/OltfJ8zBwu4jUALc72zjprwL+VkQOOz/TtUckjaFgiLFxTenHSgAfvq6c48191LYNuF0UY1JeTH+NVHUvkQAQve/xqHUFHow1r7O/E7htmv2fAz4XS7mSRedAEEiteZWm86ENZXzuuePseb2JT91+tdvFMSal2QjpODDRvz+VHysBlORksHl1IXsOnyccjqnfgjFmkVhwiAMdTs0h1R8rAXx04zLOdg7x0ps2U6sxbrLgEAc6reZwwQc3lFGem8Hjv6hzuyjGpDQLDnGgc9BqDhPSvR7+6J2r+M2ZLg432HxLxrjFgkMc6BgYxZ/mId2b2v8cuw/Us/tAPV4RMtI9/M8fHWX3gXq3i2VMSkrtv0ZxomMgaI+UovjTvWxaVcixpj6bjM8Yl1hwiAOdA6P2SGmSW9YUku718NyRZiI9pY0xV5IFhzjQaTWHKZZkpPP+dUs51drP3qMtbhfHmJRjwSEOdA6OWnCYxs2rCynPy+B//fgYvcNjbhfHmJRiwcFl42GlazBoj5Wm4fUIv3t9JZ0Do/yf50+6XRxjUooFB5d1DwUJq02dMZOK/Ew+dssqvnOgnkPnut0ujjEpw4KDyzptdPScPvWBqynNyeCvf3iUsfGw28UxJiVYcHDZhXmVMiw4zCTbn8Znt67nVGs/X3vZRk4bcyVYcHDZheDgs+Awm9vXLeWOa5fypZ/WUN855HZxjEl6Fhxc9tvpui04zGRi5PT1y/IBeGDXQb6z/5zLpTImuVlwcFnn4Chej5DhswbpueRmpnP7uqXUtA1wpLHX7eIYk9QsOLisoz9IYcCHR8TtoiSEzasLqczP5CdHm+kZCrpdHGOSlgUHl3UOjlKY7Xe7GAnDI8Ld11cwHAzx2R8fd7s4xiQtCw4u6xgIUpTtc7sYCaU8L5Nbry7hh6+d5ydHmtwujjFJyYKDyzoHRykMWHC4VO+7poTrluXx1z88SlPPsNvFMSbpWHBwWUd/kCJ7rHTJvB7hS39wPaGw8uffO0zIBscZs6AsOLhoKBhieGzc2hzmaWVRgM9uXc+BM13804tvul0cY5KKda530cQYh8JsH6Fxe2fBpZp4S9zbV+bz2Eun6R8Osa48h/s2LXe5ZMYkvphqDiJyp4icEpFaEdk5zXERkUec40dE5Ma58opIgYi8KCI1zjLf2V8oIj8XkQER+cpCXGS8mhgdbQ3Sl+dDG8qpyMvk3w410GlvjjNmQcwZHETECzwKbAHWAfeKyLpJybYAVc7PduCxGPLuBPapahWwz9kGGAH+FviL+V9WYuhwag7W5nB50r0e7rtpOSLwTHWDtT8YswBiqTncBNSqap2qBoGnga2T0mwFntSI/UCeiJTNkXcrsMtZ3wXcDaCqg6r6SyJBIqlNfMu1NofLlx/wsfW6Chq6h/nqS6fdLo4xCS+W4FABNERtNzr7YkkzW96lqtoM4CxLYi82iMh2EakWker29vZLyRo3OgedNgfryrogrluWx4bKXL60r4bXG3rcLo4xCS2W4DDdvA6TW09nShNL3nlR1SdUdaOqbiwuLl6IU15xHQOR14NmpNu8Sgtl63UVlCzx88nvvEpzr41/MGa+YgkOjcCyqO1KYPKw1JnSzJa31Xn0hLNsi73YycFGRy+8TJ+Xr92/kd7hMe7/xm/oHrT5l4yZj1iCw0GgSkRWiYgPuAfYMynNHuB+p9fSZqDXeVQ0W949wDZnfRvw7GVeS8LpHLB5lRbD+opcnrj/bZzrHOKPdh1kKBhyu0jGJJw5g4OqhoCHgOeBE8AzqnpMRHaIyA4n2V6gDqgFvgZ8cra8Tp6HgdtFpAa43dkGQETOAl8APiYijdP0jkoKnQNBa29YBLsP1HO2Y4iPbKzkcH0PW7/yK5585azbxTImocQ0CE5V9xIJANH7Ho9aV+DBWPM6+zuB22bIszKWciW6zsFRblyR73Yxkta15bncfUMFP3rtPN8/1MgfblqBx2NToxsTC5s+YwbjYeXDX/klzx1pXrTzdw0GKbY2h0X19pUF3LFuKUcae/nsT44T+R5jjJmLBYcZnO8e5khjL88ePr8o52/tGyGsUJyTsSjnN7/17quLeedVRfzrr8/y5Z/Vul0cYxKCza00gzOdgwAcONNFOKwL9jhiYj6g0+0Dkc9pH7ywzywOEeHO9aWU5Pj5wotvkh/w8V82r3C7WMbENas5zOBsRyQ49A6PcbKlf8HPHz3pnll8HhH+4fc28L5rSvi7Z9+wlwQZMwcLDjM40zFImlNb2F/XueDn7xwYJc0j5GamL/i5zfT+rbqRd1cVs7wgiz/97mE+++PjVmszZgYWHGZwpmOQtaVLWFGYtSjBoWMwSEHAh0es98yV5EvzcP/mlRQv8SJwfq0AAAwlSURBVPPt/eeo6xhwu0jGxCULDtPYfaCeo+d78YhQnO3n5ZoOvr3/3IJ+hg2Ac0+mz8vH3rGS3Mx0vvWrs/aIyZhpWHCYxnhY6RkKUpjtY1VRgOGxcVr7Fm6S2LBGurEW2QA41+RkpPOJW1dTmZfJQ7tf4+sv17ldJGPiigWHaXQPBgkrFAX8rCoKAFDXPrhg5+8dHiMUVqs5uCzLl8YfvXMVW9aX8rnnTvDZHx8nHLZxEMaABYdpRb+hLS/LR0HAx5mOhQsO1lMpfqR7PXzlvhv52C0r+eavzvDH332N4eC428UyxnUWHKbRMfGeBeeb/eqiAKfbBxgNLcwfjd8GH6s5xIPvHWygqiSbLetL2Xu0mff+40t8eV+N28UyxlUWHKbROTBKRrqHLF/kPQvXlucyGgrzy5qOBTt/uldYkmFjEOOFiPCuqmLuv3klvcNjPPpSLS8ca3G7WMa4xoLDNDoHghRl+xGnm+makgAZ6R6eO7ow8yx1DgYpDPitG2scWlu6hAffexUFWT62P3WIT//wqE35bVKSBYdpdAyMXvTIJ83j4dqyXF481rogj5Y6BoLW3hDHCgI+drxnDZ+4dTVPH6znzi++zAvHWmzSPpNSLDhMMjI2Tu/w2JT3LKyvyKV/NHTZj5bCqnQ7NQcTv9I8HlYUBHjgHasYHhtn+1OHeP8XfsHn/+Ok20Uz5oqw4DBJfdcQClO6ma4pCZCbmX7Zj5Z6hsYYV7XXgyaI1cXZ/Mn7qvidDWU09Yzw1ZdO88C/HuRAXad1ezVJzVpEJ5nosjr5j3eax8MH1i3lP95oYTQ0jj/NO6/zdzo9lWyMQ+LweoSb1xRx4/J8Xqnr5MCZLvY9sZ/y3Ay23lDBtptXUpprU6+b5GI1h0kON/Tgkem7mX5wQxn9oyH2XkbtYWKktbU5JB5/upf3rC3hz95fxUc3LmNJRjr/8ovTvOPhn/F7j/2aky19bhfRmAVjNYdJnj/WwqqiABnpU2sG764qZl1ZDv/0wpvc9dayS649qCrV57opy81gid9ufaLyp3m5flke1y/Lo3swyMu17VSf7ebOL77MDcvz+OjGZbx3bYnVJkxCs5pDlNq2AeraB1lXnjvtcY9H+PRd19DYPcxTr1z6RHyvnO6krX+UW9YUXugmaxJbfsDHh6+rYOed13DXW8to6Brm0z88yua/38fNf7+P//Zk9YV3gxiTSOzra5TnnUFP68pypj0+Mfd/VUk2//TCmwhCps/LfZuWx3T+b/7qLFk+Lxsq8xamwCZuZPnTeOdVRbxjTSGtfaOcau3nZHMfLx5v5cXjrVxTuoT3XVPCrVcXc92yvGlrpsbEEwsOUV441sJ1lblzvoDnjmtLefTntex5/Tz/6cbKmM5d3znEvpOt3Hp1Meleq7AlKxGhNDeD0twMbr26mJ6hIEfP93KypZ/Hf3Gar750Gq8IZXkZ3LKmkBWFAa4qyWZ9RS7luRlWozRxw4KDo7l3mNcbe/kfd6ydM215XibvvaaEn51so7VvlJvXFHL10iWz5vnGL+vwirB5VeFCFdkkgLwsH++qKuZdVcWMjI1zpmOQ+q4h6ruG+Pmpdtr7Gy+kzc9Kpyw3k6U5fsrzMllRmMXyggArCrNYUZhFls9+Xc2VE9P/NhG5E/gS4AW+rqoPTzouzvG7gCHgY6r66mx5RaQA+B6wEjgLfFRVu51jnwYeAMaBP1HV5y/rKmPwwrFWIFIr+M2ZrjnTv/8tS1mWn8X3X23kg4+8zN3XV/CJW9dwVUn2Rekauob4zI+P8dMTbfz+2yrJsdeCpqyMdC9vKcvhLVGPLUfHxmntH6WpZ5jm3hH6hsc41dLP/rouhscuHo2f7U+jIBCZJbgo2+es+yl09hVk+y6sFwb8ZKR7rCZi5k3mmhJARLzAm8DtQCNwELhXVY9HpbkL+GMiwWET8CVV3TRbXhH5PNClqg+LyE4gX1X/SkTWAd8FbgLKgZ8CV6vqjPNWbNy4Uaurq+d1AwZHQzx3pJkv/7wGn9fDvv/+nkt6r3D/yBhNPcN8r7qBkbEwWT4v6V4PHoHB4DhjoTDpXg+3vaWEW9YU4fXYL6uJzXBwnK7BIJ2Do3QPBukfDTE4GmIwOB5ZjoYYHhtnbHzm32GvR8hK91KQ7SPP+WISCitej+BP8+BP8+JL8zjrk7bTf7ud5hG8HsEjgscjeEUu7EvzOkuPk84bdczjQQQ8EnnkJhA5hwgiOMfkwjLy6xFZRu8XJ/+F/fx2e7rlhOjftug4KVFHpoufnqjP8sTJ72xoPExYId0riAjhsDIYDKFEXl41HyJySFU3TncslprDTUCtqtY5J3sa2Aocj0qzFXhSI5Fmv4jkiUgZkVrBTHm3Au9x8u8CXgL+ytn/tKqOAmdEpNYpwyuxXnCsqs92se2bv2EwOM6a4gB/9zvXXvI5lmSks7Y0nU/dvpZXz3UzMBoiFFZUlTSP4E/38vaVBXO2YxgzWabPS4Uvk4r8zBnTqCqjofCFYDEYHGdgNMTQaIixsBKOOj4UHL/wx3ZsPBz5vzquhMJhZ6mExsORpbNug8AjLgQKkUjEUVAUVVC4aN6t6CDIRGAkEuDCqpF7qhPretE9nvw5E3nHnH+XiTS+NA8jY2EAPnxdOY/ce8OCX3MswaECaIjabiRSO5grTcUceZeqajOAqjaLSEnUufZPc66LiMh2YLuzOSAip2K4lhmdA372FwAUAQszN7fjWwt5sitvwe9HgrP7cTG7H1Nd0XvyZeDL9807+4qZDsQSHKarU03+PjFTmljyzufzUNUngCfmONclE5HqmapZqcjux8XsflzM7sdUyXJPYulT2Qgsi9quBJpiTDNb3lbn0RPOsu0SPs8YY8wiiiU4HASqRGSViPiAe4A9k9LsAe6XiM1Ar/PIaLa8e4Btzvo24Nmo/feIiF9EVgFVwG/meX3GGGPmYc7HSqoaEpGHgOeJdEf9pqoeE5EdzvHHgb1EeirVEunK+vHZ8jqnfhh4RkQeAOqBjzh5jonIM0QarUPAg7P1VFoEC/6oKsHZ/biY3Y+L2f2YKinuyZxdWY0xxqQem8fBGGPMFBYcjDHGTGHBwSEid4rIKRGpdUZsJyURWSYiPxeREyJyTET+1NlfICIvikiNs8yPyvNp576cEpE7ova/TUSOOscekQSeq0FEvCLymoj8xNlO2fvhDGL9voicdP6f3Jzi9+PPnd+VN0TkuyKSkRL3Q1VT/odIY/lpYDXgA14H1rldrkW61jLgRmd9CZHpTdYBnwd2Ovt3Av/grK9z7ocfWOXcJ69z7DfAzUTGpvw7sMXt67uM+/IpYDfwE2c7Ze8HkRkL/quz7gPyUvV+EBmAewbIdLafAT6WCvfDag4RF6YIUdUgMDHNR9JR1WZ1JkVU1X7gBJFfgK1E/ijgLO921i9MZ6KqZ4j0SLvJGZuSo6qvaOR//pNReRKKiFQCHwS+HrU7Je+HiOQA7wa+AaCqQVXtIUXvhyMNyBSRNCCLyLirpL8fFhwiZpr+I6mJyErgBuAAk6YzAaKnM5lpapTGafYnoi8CfwmEo/al6v1YDbQD33Ies31dRAKk6P1Q1fPAPxLpbt9MZAzXC6TA/bDgEDGfaT4SmohkAz8A/kxV+2ZLOs2++U6NEndE5ENAm6oeijXLNPuS5n4Q+ZZ8I/CYqt4ADBJ5bDKTpL4fTlvCViKPiMqBgIj84WxZptmXkPfDgkNESk3ZISLpRALDd1T1h87uS53OpNFZn7w/0bwD+LCInCXyOPF9IvJtUvd+NAKNqnrA2f4+kWCRqvfj/cAZVW1X1THgh8AtpMD9sOAQEcsUIUnB6SHxDeCEqn4h6tAlTWfiVKX7RWSzc877o/IkDFX9tKpWqupKIv/uP1PVPyR170cL0CAiE69EvI3IbAUpeT+IPE7aLCJZznXcRqSdLvnvh9st4vHyQ2T6jzeJ9C74G7fLs4jX+U4i1dkjwGHn5y6gENgH1DjLgqg8f+Pcl1NE9bAANgJvOMe+gjPiPlF/iLxfZKK3UsreD+B6oNr5P/J/gfwUvx+fAU461/IUkZ5ISX8/bPoMY4wxU9hjJWOMMVNYcDDGGDOFBQdjjDFTWHAwxhgzhQUHY4wxU1hwMMYYM4UFB2OMMVP8f6I0wbEJYSOEAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "sns.distplot(np.diff(columnptr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scaling File Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _worker_part(filename, part, n_parts, rows_per_part, start_idx, stop_idx):\n",
    "    ## Divide rows into n_workers parts & load parts\n",
    "    import h5py\n",
    "    \n",
    "    print(\"Opening \" + str(filename))\n",
    "\n",
    "    f = h5py.File(filename, mode=\"r\")\n",
    "    dset = f[\"mm10\"]\n",
    "    columnptr = dset[\"indptr\"][:]\n",
    "    \n",
    "    \n",
    "    if stop_idx >= len(columnptr):\n",
    "        stop_idx = len(columnptr)-1\n",
    "    \n",
    "    start_ind = columnptr[start_idx]\n",
    "    stop_ind = columnptr[stop_idx]\n",
    "\n",
    "    inds = dset[\"indices\"]\n",
    "    with inds.astype(np.int32):\n",
    "        out = inds[start_ind:stop_ind]\n",
    "    \n",
    "    data = dset[\"data\"]\n",
    "    with data.astype(np.float32):\n",
    "        out_data = data[start_ind:stop_ind]\n",
    "        \n",
    "    ## Load into a sparse Dask array\n",
    "    rowptr = np.zeros((rows_per_part+1,), dtype=cp.int32)\n",
    "    \n",
    "    c = columnptr[start_idx:stop_idx+1]\n",
    "    \n",
    "    cumsum = np.cumsum(np.diff(c))\n",
    "    rowptr[1:] = cumsum\n",
    "    \n",
    "    print(str(rowptr[-1]))\n",
    "    \n",
    "    ret = scipy.sparse.csr_matrix((out_data, out, rowptr), \n",
    "                                   shape=(rows_per_part, dset[\"shape\"][0]),\n",
    "                                  dtype=cp.float32)\n",
    "    return cp.sparse.csr_matrix(ret, dtype=cp.float32)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 15.3 ms, sys: 855 µs, total: 16.1 ms\n",
      "Wall time: 14.4 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import dask\n",
    "import dask.array as da\n",
    "import math\n",
    "\n",
    "nnz = 2624828308\n",
    "nnz = 251544125\n",
    "\n",
    "load_part = dask.delayed(_worker_part)\n",
    "\n",
    "n_parts = 10\n",
    "\n",
    "\n",
    "def create_dask_arrays(dset, filename, n_parts):\n",
    "\n",
    "    final_shape = tuple(dset[\"shape\"][:][::-1])\n",
    "    \n",
    "    total_rows = dset[\"shape\"][1]\n",
    "    rows_per_part = int(total_rows / n_parts)\n",
    "\n",
    "    out_arrs = []\n",
    "    for part in range(n_parts):\n",
    "\n",
    "        if rows_per_part * part > total_rows:\n",
    "            rows_per_part = total_rows - (rows_per_part * part)\n",
    "\n",
    "        start_idx = part * rows_per_part\n",
    "        stop_idx = start_idx + rows_per_part\n",
    "        out_arrs.append(da.from_delayed(load_part(filename, part, n_parts, rows_per_part, start_idx, stop_idx),\n",
    "                                        dtype=cp.float32, \n",
    "                                        shape=(rows_per_part, final_shape[1])))\n",
    "    return out_arrs\n",
    "\n",
    "\n",
    "final_arr = da.concatenate(create_dask_arrays(dset, one_mil_file, n_parts))\n",
    "final_arr = final_arr.persist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def conv_to_cp_csr(x):\n",
    "    \n",
    "    print(str(type(x)))\n",
    "    indptr = cp.array(x.indptr)\n",
    "    indices = cp.array(x.indices)\n",
    "    data = cp.array(x.data)\n",
    "    \n",
    "    return cp.sparse.csr_matrix((data, indices, indptr))\n",
    "\n",
    "# final_arr = final_arr.map_blocks(conv_to_cp_csr, meta=cp.sparse.csr_matrix(cp.zeros(1))).astype(cp.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_arr = final_arr.map_blocks(lambda x: cp.sparse.csr_matrix(rapids_scanpy_funcs.filter_cells(x, min_genes=200, max_genes=6000)), \n",
    "                                 meta=cp.sparse.csr_matrix(cp.zeros(1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "final_arr = final_arr.persist()\n",
    "final_arr.compute_chunk_sizes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cupyx.scipy.sparse.csr.csr_matrix"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(final_arr[0].compute())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perf_sum(x):\n",
    "    print(str(type(x)))\n",
    "    return cp.sparse.csr_matrix(x.get()).sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dask.array as da"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_counts = final_arr.map_blocks(perf_sum, meta=cp.sparse.csr_matrix(cp.zeros(1))).persist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<tr>\n",
       "<td>\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr><td> </td><th> Array </th><th> Chunk </th></tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr><th> Bytes </th><td> 289.24 GB </td> <td> 29.05 GB </td></tr>\n",
       "    <tr><th> Shape </th><td> (1291330, 27998) </td> <td> (129676, 27998) </td></tr>\n",
       "    <tr><th> Count </th><td> 10 Tasks </td><td> 10 Chunks </td></tr>\n",
       "    <tr><th> Type </th><td> float64 </td><td> numpy.ndarray </td></tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</td>\n",
       "<td>\n",
       "<svg width=\"79\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"29\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"29\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"29\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"36\" x2=\"29\" y2=\"36\" />\n",
       "  <line x1=\"0\" y1=\"48\" x2=\"29\" y2=\"48\" />\n",
       "  <line x1=\"0\" y1=\"60\" x2=\"29\" y2=\"60\" />\n",
       "  <line x1=\"0\" y1=\"72\" x2=\"29\" y2=\"72\" />\n",
       "  <line x1=\"0\" y1=\"84\" x2=\"29\" y2=\"84\" />\n",
       "  <line x1=\"0\" y1=\"96\" x2=\"29\" y2=\"96\" />\n",
       "  <line x1=\"0\" y1=\"108\" x2=\"29\" y2=\"108\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"29\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"29\" y1=\"0\" x2=\"29\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.000000,0.000000 29.488329,0.000000 29.488329,120.000000 0.000000,120.000000\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"14.744165\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >27998</text>\n",
       "  <text x=\"49.488329\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,49.488329,60.000000)\">1291330</text>\n",
       "</svg>\n",
       "</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "dask.array<perf_sum, shape=(1291330, 27998), dtype=float64, chunksize=(129676, 27998), chunktype=numpy.ndarray>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_counts = gene_counts.persist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 33.6 ms, sys: 8.09 ms, total: 41.7 ms\n",
      "Wall time: 39.3 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "gene_counts = gene_counts.map_blocks(lambda x: x.get(), meta=cp.sparse.csr_matrix(cp.zeros(1))).compute().sum(axis=0) > 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _filter_genes(sparse_gpu_array, thr):\n",
    "    return sparse_gpu_array[:,thr]\n",
    "\n",
    "final_arr = final_arr.map_blocks(lambda x: cp.sparse.csr_matrix(_filter_genes(x.get(), gene_counts)), meta=cp.sparse.csr_matrix(cp.zeros(1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_arr = final_arr.persist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 382 ms, sys: 51.5 ms, total: 434 ms\n",
      "Wall time: 5.41 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<tr>\n",
       "<td>\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr><td> </td><th> Array </th><th> Chunk </th></tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr><th> Bytes </th><td> 247.96 GB </td> <td> 24.90 GB </td></tr>\n",
       "    <tr><th> Shape </th><td> (1291330, 24002) </td> <td> (129676, 24002) </td></tr>\n",
       "    <tr><th> Count </th><td> 10 Tasks </td><td> 10 Chunks </td></tr>\n",
       "    <tr><th> Type </th><td> float64 </td><td> numpy.ndarray </td></tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</td>\n",
       "<td>\n",
       "<svg width=\"78\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"28\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"28\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"28\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"36\" x2=\"28\" y2=\"36\" />\n",
       "  <line x1=\"0\" y1=\"48\" x2=\"28\" y2=\"48\" />\n",
       "  <line x1=\"0\" y1=\"60\" x2=\"28\" y2=\"60\" />\n",
       "  <line x1=\"0\" y1=\"72\" x2=\"28\" y2=\"72\" />\n",
       "  <line x1=\"0\" y1=\"84\" x2=\"28\" y2=\"84\" />\n",
       "  <line x1=\"0\" y1=\"96\" x2=\"28\" y2=\"96\" />\n",
       "  <line x1=\"0\" y1=\"108\" x2=\"28\" y2=\"108\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"28\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"28\" y1=\"0\" x2=\"28\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.000000,0.000000 28.621212,0.000000 28.621212,120.000000 0.000000,120.000000\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"14.310606\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >24002</text>\n",
       "  <text x=\"48.621212\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,48.621212,60.000000)\">1291330</text>\n",
       "</svg>\n",
       "</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "dask.array<lambda, shape=(1291330, 24002), dtype=float64, chunksize=(129676, 24002), chunktype=numpy.ndarray>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "final_arr.compute_chunk_sizes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_arr = final_arr.map_blocks(lambda x:  rapids_scanpy_funcs.normalize_total(x.get(), target_sum=1e4), meta=cp.sparse.csr_matrix(cp.zeros(1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<tr>\n",
       "<td>\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr><td> </td><th> Array </th><th> Chunk </th></tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr><th> Bytes </th><td> 247.96 GB </td> <td> 24.90 GB </td></tr>\n",
       "    <tr><th> Shape </th><td> (1291330, 24002) </td> <td> (129676, 24002) </td></tr>\n",
       "    <tr><th> Count </th><td> 20 Tasks </td><td> 10 Chunks </td></tr>\n",
       "    <tr><th> Type </th><td> float64 </td><td> numpy.ndarray </td></tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</td>\n",
       "<td>\n",
       "<svg width=\"78\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"28\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"28\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"28\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"36\" x2=\"28\" y2=\"36\" />\n",
       "  <line x1=\"0\" y1=\"48\" x2=\"28\" y2=\"48\" />\n",
       "  <line x1=\"0\" y1=\"60\" x2=\"28\" y2=\"60\" />\n",
       "  <line x1=\"0\" y1=\"72\" x2=\"28\" y2=\"72\" />\n",
       "  <line x1=\"0\" y1=\"84\" x2=\"28\" y2=\"84\" />\n",
       "  <line x1=\"0\" y1=\"96\" x2=\"28\" y2=\"96\" />\n",
       "  <line x1=\"0\" y1=\"108\" x2=\"28\" y2=\"108\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"28\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"28\" y1=\"0\" x2=\"28\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.000000,0.000000 28.621212,0.000000 28.621212,120.000000 0.000000,120.000000\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"14.310606\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >24002</text>\n",
       "  <text x=\"48.621212\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,48.621212,60.000000)\">1291330</text>\n",
       "</svg>\n",
       "</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "dask.array<lambda, shape=(1291330, 24002), dtype=float64, chunksize=(129676, 24002), chunktype=numpy.ndarray>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "final_arr.compute_chunk_sizes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_arr = final_arr.map_blocks(lambda x: x.log1p(), meta=cp.sparse.csr_matrix(cp.zeros(1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.8 s, sys: 275 ms, total: 2.08 s\n",
      "Wall time: 32.6 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<tr>\n",
       "<td>\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr><td> </td><th> Array </th><th> Chunk </th></tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr><th> Bytes </th><td> 247.96 GB </td> <td> 24.90 GB </td></tr>\n",
       "    <tr><th> Shape </th><td> (1291330, 24002) </td> <td> (129676, 24002) </td></tr>\n",
       "    <tr><th> Count </th><td> 20 Tasks </td><td> 10 Chunks </td></tr>\n",
       "    <tr><th> Type </th><td> float64 </td><td> numpy.ndarray </td></tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</td>\n",
       "<td>\n",
       "<svg width=\"78\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"28\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"28\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"28\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"36\" x2=\"28\" y2=\"36\" />\n",
       "  <line x1=\"0\" y1=\"48\" x2=\"28\" y2=\"48\" />\n",
       "  <line x1=\"0\" y1=\"60\" x2=\"28\" y2=\"60\" />\n",
       "  <line x1=\"0\" y1=\"72\" x2=\"28\" y2=\"72\" />\n",
       "  <line x1=\"0\" y1=\"84\" x2=\"28\" y2=\"84\" />\n",
       "  <line x1=\"0\" y1=\"96\" x2=\"28\" y2=\"96\" />\n",
       "  <line x1=\"0\" y1=\"108\" x2=\"28\" y2=\"108\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"28\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"28\" y1=\"0\" x2=\"28\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.000000,0.000000 28.621212,0.000000 28.621212,120.000000 0.000000,120.000000\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"14.310606\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >24002</text>\n",
       "  <text x=\"48.621212\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,48.621212,60.000000)\">1291330</text>\n",
       "</svg>\n",
       "</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "dask.array<lambda, shape=(1291330, 24002), dtype=float64, chunksize=(129676, 24002), chunktype=numpy.ndarray>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "final_arr.compute_chunk_sizes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'tcp://127.0.0.1:34593': (\"('lambda-83c45491caf52575a14faaa3910091b7', 0, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 0, 0)\",\n",
       "  \"('lambda-2cd89a3ca607d5c326ec8f5f61c745a0', 0, 0)\",\n",
       "  \"('lambda-2cd89a3ca607d5c326ec8f5f61c745a0', 8, 0)\",\n",
       "  \"('lambda-83c45491caf52575a14faaa3910091b7', 9, 0)\",\n",
       "  \"('perf_sum-1fc4a9061d0f97d16e4454eaaa973098', 0, 0)\",\n",
       "  \"('lambda-83c45491caf52575a14faaa3910091b7', 8, 0)\"),\n",
       " 'tcp://127.0.0.1:36589': (\"('perf_sum-1fc4a9061d0f97d16e4454eaaa973098', 1, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 1, 0)\",\n",
       "  \"('lambda-83c45491caf52575a14faaa3910091b7', 8, 0)\",\n",
       "  \"('lambda-83c45491caf52575a14faaa3910091b7', 1, 0)\",\n",
       "  \"('lambda-2cd89a3ca607d5c326ec8f5f61c745a0', 1, 0)\"),\n",
       " 'tcp://127.0.0.1:37667': (\"('perf_sum-1fc4a9061d0f97d16e4454eaaa973098', 2, 0)\",\n",
       "  \"('lambda-83c45491caf52575a14faaa3910091b7', 2, 0)\",\n",
       "  \"('lambda-2cd89a3ca607d5c326ec8f5f61c745a0', 2, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 2, 0)\"),\n",
       " 'tcp://127.0.0.1:38491': (\"('lambda-2cd89a3ca607d5c326ec8f5f61c745a0', 3, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 3, 0)\",\n",
       "  \"('lambda-83c45491caf52575a14faaa3910091b7', 3, 0)\",\n",
       "  \"('perf_sum-1fc4a9061d0f97d16e4454eaaa973098', 3, 0)\"),\n",
       " 'tcp://127.0.0.1:40523': (\"('lambda-83c45491caf52575a14faaa3910091b7', 4, 0)\",\n",
       "  \"('perf_sum-1fc4a9061d0f97d16e4454eaaa973098', 4, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 4, 0)\",\n",
       "  \"('lambda-2cd89a3ca607d5c326ec8f5f61c745a0', 4, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 8, 0)\",\n",
       "  \"('perf_sum-1fc4a9061d0f97d16e4454eaaa973098', 8, 0)\",\n",
       "  \"('lambda-83c45491caf52575a14faaa3910091b7', 8, 0)\"),\n",
       " 'tcp://127.0.0.1:43049': (\"('lambda-83c45491caf52575a14faaa3910091b7', 5, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 5, 0)\",\n",
       "  \"('perf_sum-1fc4a9061d0f97d16e4454eaaa973098', 5, 0)\",\n",
       "  \"('lambda-2cd89a3ca607d5c326ec8f5f61c745a0', 5, 0)\"),\n",
       " 'tcp://127.0.0.1:46539': (\"('lambda-2cd89a3ca607d5c326ec8f5f61c745a0', 6, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 6, 0)\",\n",
       "  \"('lambda-83c45491caf52575a14faaa3910091b7', 6, 0)\",\n",
       "  \"('perf_sum-1fc4a9061d0f97d16e4454eaaa973098', 6, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 7, 0)\"),\n",
       " 'tcp://127.0.0.1:46555': (\"('perf_sum-1fc4a9061d0f97d16e4454eaaa973098', 9, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 9, 0)\",\n",
       "  \"('lambda-2cd89a3ca607d5c326ec8f5f61c745a0', 9, 0)\",\n",
       "  \"('lambda-83c45491caf52575a14faaa3910091b7', 7, 0)\",\n",
       "  \"('lambda-83c45491caf52575a14faaa3910091b7', 9, 0)\",\n",
       "  \"('lambda-dd81996c1b37445a11a06b9523e9cbbd', 7, 0)\",\n",
       "  \"('lambda-2cd89a3ca607d5c326ec8f5f61c745a0', 7, 0)\",\n",
       "  \"('perf_sum-1fc4a9061d0f97d16e4454eaaa973098', 7, 0)\")}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "client.has_what()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We maintain the index of unique cells and genes in our dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "cells = cudf.Series(adata.obs_names)\n",
    "genes = cudf.Series(adata.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rows = cp.array(adata.X.indptr)\n",
    "cols = cp.array(adata.X.indices)\n",
    "vals = cp.array(adata.X.data)\n",
    "sparse_gpu_array = cp.sparse.csr_matrix((vals, cols, rows), shape=adata.X.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Verify the shape of the resulting sparse matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparse_gpu_array.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And the number of non-zero values n the matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparse_gpu_array.nnz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_load_time = time.time()\n",
    "print(\"Total data load and format time: %s\" % (data_load_time-data_load_start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "preprocess_start = time.time()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We filter the count matrix to remove cells with an extreme number of genes expressed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "filtered = rapids_scanpy_funcs.filter_cells(sparse_gpu_array, \n",
    "                                            min_genes=min_genes_per_cell, \n",
    "                                            max_genes=max_genes_per_cell,\n",
    "                                            rows_per_batch=20000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some genes will now have zero expression in all cells. We filter out such genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "filtered, genes = rapids_scanpy_funcs.filter_genes(filtered, genes, min_cells=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The size of our count matrix is now reduced."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalize"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We normalize the count matrix so that the total counts in each cell sum to 1e4."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "normalized = rapids_scanpy_funcs.normalize_total(filtered, target_sum=1e4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we log transform the count matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "normalized = normalized.log1p()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select Most Variable Genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We convert the count matrix to an annData object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "adata = anndata.AnnData(normalized.get())\n",
    "adata.var_names = genes.to_pandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using scanpy, we filter the count matrix to retain only the 5000 most variable genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.pp.highly_variable_genes(adata, n_top_genes=10, flavor=\"cell_ranger\")\n",
    "adata = adata[:, adata.var.highly_variable]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Regress out confounding factors (number of counts, ribosomal gene expression)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now perform regression on the count matrix to correct for confounding factors -  for example purposes, we use the number of counts and the expression of ribosomal genes. Many workflows use the expression of mitochondrial genes (named starting with `MT-`).\n",
    "\n",
    "Before regression, we save the 'raw' expression values of the ACE2 and TMPRSS2 genes to use for labeling cells afterward. We will also store the expression of n epithelial marker gene (EPCAM)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes[genes == \"ACE2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "tmp_norm = normalized.tocsc()\n",
    "ACE2_raw = tmp_norm[:, genes[genes == \"ACE2\"].index[0]].todense().ravel()\n",
    "TMPRSS2_raw = tmp_norm[:, genes[genes == \"TMPRSS2\"].index[0]].todense().ravel()\n",
    "EPCAM_raw = tmp_norm[:, genes[genes == \"EPCAM\"].index[0]].todense().ravel()\n",
    "\n",
    "del tmp_norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = adata.var_names\n",
    "ribo_genes = adata.var_names.str.startswith(RIBO_GENE_PREFIX)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "filtered = adata.X"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now calculate the total counts and the percentage of ribosomal counts for each cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "n_counts = filtered.sum(axis=1)\n",
    "percent_ribo = (filtered[:,ribo_genes].sum(axis=1) / n_counts).ravel()\n",
    "\n",
    "n_counts = cp.array(n_counts).ravel()\n",
    "percent_ribo = cp.array(percent_ribo).ravel()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And perform regression:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "normalized = cp.array(adata.X.todense(), order=\"F\")\n",
    "normalized = rapids_scanpy_funcs.regress_out(normalized, n_counts, percent_ribo)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scale"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we scale the count matrix to obtain a z-score and apply a cutoff value of 10 standard deviations, obtaining the preprocessed count matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "normalized = rapids_scanpy_funcs.scale(normalized, max_value=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "preprocess_time = time.time()\n",
    "print(\"Total Preprocessing time: %s\" % (preprocess_time-preprocess_start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cluster & Visualize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_start = time.time()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We store the preprocessed count matrix as an AnnData object, which is currently in host memory. We also add the expression levels of the marker genes as observations to the annData object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "var_names = adata.var_names\n",
    "adata = anndata.AnnData(normalized.get())\n",
    "adata.var_names = var_names\n",
    "adata.obs[\"ACE2_raw\"] = ACE2_raw.get()\n",
    "adata.obs[\"TMPRSS2_raw\"] = TMPRSS2_raw.get()\n",
    "adata.obs[\"EPCAM_raw\"] = EPCAM_raw.get()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reduce"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use PCA to reduce the dimensionality of the matrix to its top 50 principal components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "adata.obsm[\"X_pca\"] = PCA(n_components=n_components).fit_transform(adata.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm[\"X_pca\"].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TSNE + K-means"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We next use t-SNE on the top 20 principal components to visualize the cells in two dimensions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "adata.obsm['X_tsne'] = TSNE().fit_transform(adata.obsm[\"X_pca\"][:,:tsne_n_pcs])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We cluster the cells using k-means on the principal components. For example purposes, we set k=20."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "kmeans = KMeans(n_clusters=k, random_state=0).fit(adata.obsm['X_pca'])\n",
    "adata.obs['kmeans'] = kmeans.labels_.astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We visualize the cells using t-SNE and label cells by color according to the k-means clustering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.pl.tsne(adata, color=[\"kmeans\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Separately, we label the cells using the 'raw' expression of the EPCAM, ACE2 and TMPRSS2 marker genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.pl.tsne(adata, size=4, color=[\"ACE2_raw\"], color_map=\"Blues\", vmax=1, vmin=-0.05)\n",
    "sc.pl.tsne(adata, size=4, color=[\"TMPRSS2_raw\"], color_map=\"Blues\", vmax=1, vmin=-0.05)\n",
    "sc.pl.tsne(adata, size=4, color=[\"EPCAM_raw\"], color_map=\"Reds\", vmax=1,  vmin=-0.05)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### UMAP + Louvain"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also visualize the cells using the UMAP algorithm in Rapids. Before UMAP, we need to construct a k-nearest neighbors graph in which each cell is connected to its nearest neighbors. This can be done conveniently using rapids functionality already integrated into Scanpy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.pp.neighbors(adata, n_neighbors=n_neighbors, n_pcs=knn_n_pcs, method='rapids')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The UMAP function from Rapids is also integrated into Scanpy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.tl.umap(adata, min_dist=umap_min_dist, spread=umap_spread, method='rapids')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we use the Louvain algorithm for graph-based clustering, once again using the `rapids` option in Scanpy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.tl.louvain(adata, resolution=louvain_resolution, flavor='rapids')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We plot the cells using the UMAP visualization, and using the Louvain clusters as labels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.pl.umap(adata, color=[\"louvain\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also view cells using UMAP and labeling by raw EPCAM, ACE2 and TMPRSS2 expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.pl.umap(adata, size=4,color=[\"ACE2_raw\"], color_map=\"Blues\", vmax=1, vmin=-0.05)\n",
    "sc.pl.umap(adata, size=4, color=[\"TMPRSS2_raw\"], color_map=\"Blues\", vmax=1, vmin=-0.05)\n",
    "sc.pl.umap(adata, size=4, color=[\"EPCAM_raw\"], color_map=\"Reds\", vmax=1, vmin=-0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_time = time.time()\n",
    "print(\"Total cluster time : %s\" % (cluster_time-cluster_start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential expression analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once we have done clustering, we can compute a ranking for the highly differential genes in each cluster. Here we use the Louvain clusters as labels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_labels = cudf.Series.from_categorical(adata.obs[\"louvain\"].cat)\n",
    "var_names = cudf.Series(var_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use logistic regression to identify the top 50 genes distinguishing each cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "scores, names, reference = rapids_scanpy_funcs.rank_genes_groups(\n",
    "    normalized, \n",
    "    cluster_labels, \n",
    "    var_names, \n",
    "    n_genes=ranking_n_top_genes, groups='all', reference='rest')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "adata.uns[\"rank_genes_groups\"] = {}\n",
    "adata.uns[\"rank_genes_groups\"][\"params\"] = dict(groupby=\"louvain\", method=\"logreg\", reference=reference, use_raw=False)\n",
    "adata.uns[\"rank_genes_groups\"]['scores'] = scores\n",
    "adata.uns[\"rank_genes_groups\"]['names'] = names\n",
    "sc.pl.rank_genes_groups(adata, n_genes=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Zoomed View"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The speedup offered by Rapids makes it easy to interactively re-analyze subsets of cells. To illustrate this, we select epithelial cells (EPCAM+) from the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reanalysis_start = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "epcam_cells = EPCAM_raw > 0.0\n",
    "\n",
    "normalized_epcam_cells = normalized[epcam_cells]\n",
    "ACE2_raw_epcam = ACE2_raw[epcam_cells]\n",
    "TMPRSS2_raw_epcam = TMPRSS2_raw[epcam_cells]\n",
    "\n",
    "var_names = adata.var_names\n",
    "adata = anndata.AnnData(normalized_epcam_cells.get())\n",
    "print(adata.X.shape)\n",
    "adata.var_names = var_names\n",
    "adata.obs[\"ACE2_raw\"] = ACE2_raw_epcam.get()\n",
    "adata.obs[\"TMPRSS2_raw\"] = TMPRSS2_raw_epcam.get()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can repeat the dimension reduction, clustering and visualization using this subset of cells in seconds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "adata.obsm[\"X_pca\"] = PCA(n_components=n_components).fit_transform(adata.X)\n",
    "sc.pp.neighbors(adata, n_neighbors=n_neighbors, n_pcs=knn_n_pcs, method='rapids')\n",
    "sc.tl.umap(adata, min_dist=umap_min_dist, spread=umap_spread, method='rapids')\n",
    "sc.tl.louvain(adata, flavor='rapids')\n",
    "\n",
    "sc.pl.umap(adata, color=[\"louvain\"])\n",
    "sc.pl.umap(adata, color=[\"ACE2_raw\"], color_map=\"Blues\", vmax=1, vmin=-0.05)\n",
    "sc.pl.umap(adata, color=[\"TMPRSS2_raw\"], color_map=\"Blues\", vmax=1, vmin=-0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reanalysis_time = time.time()\n",
    "print(\"Total reanalysis time : %s\" % (reanalysis_time-reanalysis_start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Full time: %s\" % (time.time() - start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "client.close()\n",
    "cluster.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (genomics2)",
   "language": "python",
   "name": "genomics2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
